### ISSUE TITLE
Raised on behalf of @RaymondVA:   Various i2 Connect/Python related issues.

### ISSUE THREAD

**Issue by Anthony-Johnson-i2** [i2group/analyze-connect#18]
Raymond says:

Hi,
I tried to execute pipenv run python app.py in cmd in the nypd-connector file path. The link reference is here: 

[https://i2group.github.io/analyze-connect/content/example-connectors/run-python.html]

But an error occurs: File "app.py", linke 25, in from flask import Flask ModuleNotFoundError: No module named 'flask'

Also, in the project directory (where the Pipfile is located), I tried runing the setup executable file but I don't see any changes to the folder, is this normal if I'm using the sample solution?

I also am unable to get the connector services "tabs" to appear in ANB Premium although I am using the sample solution.

**Comment by Anthony-Johnson-i2** [comment]
Hi Raymond

When you  run the `setup.bat ` it should create a new file called `Pipfile.lock` in the folder where you run.

When I run
 `pipenv run python app.py`  
  in the same folder it starts the flask app up fine and tells me that it is now running on port 9081 so it could be that Flask is not installed in your environment or not installed for the python you are now running with. 
For me it got installed as part of my python install.

As for the sample connector services tabs that you can not see,  in most cases they  will not be present when you click the External Searches button in ANBP if either the connector was not defined in the topology as expected when you deployed,  or if  you have not started the connector  before you start liberty.

Cheers

**Comment by RaymondVA** [comment]
Hi Anthony,

On Windows, when I tried to run the **setup.bat**
The **Pipfile.lock** file that you mentioned does not appear, it simply does nothing except show a glimpse of the cmd before it disappears. The setup.bat file is 1KB so I worry it is missing some execution code. Is the purpose of the setup just to create the Pipfile.lock file and is this file specific to your machine (i.e if you shared the file with me, would there be problems?)

When I run **pipenv run python app.py** in the same folder, it is still coming up with the error "ModuleNotFoundError: No module named 'flask'" although I'm certain I have flask installed. 
![image](https://user-images.githubusercontent.com/101074439/179867546-624585d0-6999-4d75-94b7-46110ccd25fd.png)

Reason is because VS Code is able to recognise import flask, there was no errors raised when I entered import flask in the Python app itself, so I'm also confused why it is not recognizing this. I'm unsure how you have installed python with flask (the python custom installer didn't show flask as an option) because I have had to manually install flask using the command **pip install Flask** in cmd. Maybe if you can tell me how it is done along with the Python installer, flask will work for me?

For the connector services tabs, I have defined them in the topology file and the services tabs still didn't appear, for example:
![image](https://user-images.githubusercontent.com/101074439/179867479-34fa5430-c14c-41da-a690-62266fbff75a.png)
The following commands I have executed:
setup -t generateDefaults (only once to generate defaults in toolkit)
setup -t deploy
setup -t ensureExampleUserRegistry
setup -t start

I did not use start liberty as it seemed to do the exact same thing as setup -t start. If there is any difference, please let me know and I'll start using the liberty commands.

**Comment by Anthony-Johnson-i2** [comment]
Hi Raymond.

**1:** The toolkit `startLiberty ` command only starts Liberty and is designed to be used in  a deployment where you have Solr Zookeeper and Liberty on different host machines. The toolkit `start` command starts Liberty and Solr and Zookeeper  and is designed to be used in a deployment where they are all running on the same host machine.

**2:** Defining the connectors in the topology as you have done is correct but unless you have successfully  started up the application servers for those connectors before you start  the Liberty application server  you will not see any of the search services  configured for them. 

For the example-connector that would mean running `npm start` in its folder in the toolkit  and for the nypd-connector, in your case that would mean successfully starting the Flask application server  using  `pipenv run python app.py`  in its folder. 

Until they are up and running correctly and you have logged in, ANBP will not be given the information about the services they support so can not display them.

**3:** I manually downloaded Python from here :

![image](https://user-images.githubusercontent.com/98407720/179957974-f662be22-bd0a-472e-8bef-c6c9f5e4f6a7.png)

I ran that installer `python-3.10.5-amd64.exe `as an Administrator and  ticked the boxes  so it would add Python 3.1.0 to PATH and install the launcher for all users.

I chose 'Customize installation' but only so I could alter the install path if needed.
I left all optional features ticked  as they are by default  (there is no mention of Flask).

On the  advanced options page I again ticked everything (probably not needed) and altered the install location 

I then hit the install button:

When it displayed the success dialog  I disabled the path length limit.

Typing `python -V `in a command window showed me I have Python 3.10.5  installed

So I was mistaken  about Flask being installed by the Python installer as I saw no option for it.

**4:** I then ran `setup .bat` (mine is only 1k like yours)  and it then installs all of the dependencies (**Flask as well** if it is not already present) .

If I look in the  `Pipfile`,  flask is mentioned in there as a dependency so it will get installed when your `setup.bat ` runs correctly  (unless it is found to already exist).

In my clean VM deployment  setup.bat did not run  **unless I already had git for windows installed**. Until I had that , it said it could not find a path.

I expect that as you do not get the `Pipfile.lock `created your setup.bat is failing to get the  dependencies downloaded and  locked in for this connector. (**Do you have git for windows installed?**)

If I try and run my connector by just saying `python app.py ` I get the same error as you do 

![image](https://user-images.githubusercontent.com/98407720/179964888-a755cd82-7cdc-41ba-a5c4-b284c6aae77a.png)

this is because after running `setup.bat` Flask installed in my case for the pipenv virtual environment and if I do not use that, it can not be seen.

If I run the command as `pipenv run python app.py ` then it all fires up correctly as I am then in the environment that Flask has been installed for.

I expect in your case , you have Flask installed but for different virtual environment (perhaps a VSCode virtual environment)  or perhaps you have more than one version of Python installed and Flask got set up for the default one but that is not now the one you are using.


Cheers

**Comment by RaymondVA** [comment]
Hi Anthony,

This is very helpful. I managed to run the **python app.py** command successfully after installing **git for windows**. So now I can see:
![image](https://user-images.githubusercontent.com/101074439/180138940-f8752200-96a3-4e9e-9b2b-07f51bd8c955.png)


Unfortunately, the python nypd services tabs still does not appear on ANB, but the example connector services do appear. I have made sure to stop the analyze server and rerun the multiple commands (setup -t ...), but the python services still don't appear.
I think this is not a problem of running the connector, because the services tab should still appear even if the services don't return any data.

Also, can you help me confirm that I use the App Token or the Secret Token.
![image](https://user-images.githubusercontent.com/101074439/180140439-86b8d22e-2a72-4011-81bb-a39bc576bc56.png)


Thank you for your help!

**Comment by Anthony-Johnson-i2** [comment]
Hi Raymond, glad to see you are making progress and can now start up the NYPD connector.

I just had another look at the image you showed me for your Topology file connector settings and I can see an issue with it and  typo.

The entries in the topology file for the connectors tell the Liberty gateway what connectors  to monitor for and where to look for them.

You have defined that both the example-connector and the nypd-connector are to be accessed on  localhost, which is fine, but you have given them both port 9081. 

This is the correct port for the nypd-connector but is not the one the one the example-connector would  be running on by default. 

You also have named the NYPD Connector  NY**DP** by mistake as I have also highlighted in the image  below.

![image](https://user-images.githubusercontent.com/98407720/180143337-592deb88-304e-4286-bbcd-553ee7108d73.png)

If you have not altered the configuration for the example connector then when you run `npm start`  it will run the node application  on port 3700 you can check this as when you start it it tells you what port it is listening on.

![image](https://user-images.githubusercontent.com/98407720/180144817-5f37c4cc-a0ab-41c2-a0e8-ae0c5b6fc19c.png)

If you fix your topology file and  run `setup -t deploy`   you should get further.

Also don't forget that we only check for the live connectors when we start up  so you need to have both of your connectors running before you  execute teh setup -t  start toolkit command.

I have not played with the NYPD connector much so will have to find out about the API keys question for you.

Cheers

**Comment by Anthony-Johnson-i2** [comment]
Hi Raymond 

I spoke with one of my colleagues, as I was unable to set up an account to try myself  this morning, and she said that it is the App Token you need to copy. 

She also kindly offered to update the documentation to make that clearer.

Cheers

**Comment by RaymondVA** [comment]
Hi Anthony,

Thank you for getting back to me :)

In my topology file, I have fixed the spelling, I had originally thought the "Name" is just to give a label for the user to see so it doesn't affect the i2analyze/code or anything; please correct me if I'm wrong.

I have changed the port for the example connector to 3700 and kept the port for NYPD to 9081. However, when I do this, i2analyze server is unable to start due to an error. Here is the error log if you wish to view it.
[console.log](https://github.com/i2group/analyze-connect/files/9163709/console.log)

^i2analyze server only runs if the port numbers are the same for both connectors (Example and NYPD = 3700 or 9081).
When experimenting this, I found out if I do this ^, only 1 of the connector (and its services) will be visible. So I wonder if it is possible to have more than 1 connector and their services to appear in ANBP.

Although I was finally able to see the NYPD connector services appear in ANBP (it is highly likely the example connector was causing NYPD to not appear), the services did not work. ANBP says that "The connector returned data that is not valid. Contact your system administrator" 

Also, I'm unsure about the port numbers - what are the available port numbers and how do I know which one to use? From your answer, it sounded as though the Example connector had to be run on port 3700.

**Comment by Anthony-Johnson-i2** [comment]
Hi Raymond
As there are quite a few points here I will number them.

**1:** You are correct the mis-spelled name was just mentioned as it was incorrect.

**2:** The connectors should be running on the ports they are set up for (3700 for the example one and 9081 for  NYPD) , and yes you can have many available in ANBP at the same time. I currently have both the example connector and the NYPD connector running in my system.

The connector code defines what port they will be running on and they should be different to each other.
For the example connector that is set in app.js

![image](https://user-images.githubusercontent.com/98407720/180713447-a6c617dc-f800-4c09-8d74-a85b34423e26.png)
and when you start it up it confirms this to you:

![image](https://user-images.githubusercontent.com/98407720/180717143-7ad657c5-a487-4399-8d70-b6403318d75a.png)


For the NYPD connector you are trying it is set in app.py

![image](https://user-images.githubusercontent.com/98407720/180714043-074948a6-4ff5-40fa-a36e-3b1b9282315b.png)

again,  when you run it it confirms this to you:

![image](https://user-images.githubusercontent.com/98407720/180717388-82e09c37-1f90-410d-afd0-84681162253d.png)

**3:** The error you get from the NYPD connector actually shows it is all running as expected at the moment.
In issue 17 where you first posted your questions I mentioned this to you.
 Afira  one of our developers created that issue  and it says that ...

_An update to the [NYPD Complaint Data] has caused the sample solution search services on the [NYPD connector] to fail on execution and is only affecting the solutions written in Java and Python. The Java connector will throw a parsing exception as it tries to parse (null) to java.time.LocalTime and the Python connector will fail on server validation as it tries to map a string value to the logical type TIME as defined by the Complaint End Time property._ 


**4:** When you have both of your connectors started before you start our main Liberty/gateway server and have them on their own separate ports as they should be, then they will be queried to get their configuration information so that we have what we need to work with them.

Looking in the console.log file you provided ( We also log application and connector specific information to i2_Anlaysis_Repository.log so you can check there too if you have problems)  I can see it saying :

![image](https://user-images.githubusercontent.com/98407720/180721175-24c519ed-81be-4694-b072-971d24e05d9f.png)

So I expect it has always had this issue in your environment,  but until you were referencing the correct port for the example connector it was not getting that far. 

I think that without the example-connector issue, your NYPD connector is working to the extent that you can see its services and run searches and get the error that we expect to see because of the NYPD complaint data change. So that is as far as you can take that one for now.

5: The next stage then will be to try and find out why, in your case, when you have correctly defined in the  topology that the example connector is using  port 3700  it  is being reported as not having a default Labelling Scheme. 

Cheers

**Comment by RaymondVA** [comment]
Hi Anthony,

Thank you for getting back! 
2. It is useful to know where to define the port numbers. I was just wondering how I can know what the available port numbers I can use are. For example, can I set the port numbers to be any number (for example: 1), or does it have to be any 4-digit number, or is there a list of port numbers that I must choose from and how can I see this.

5. About the default Labelling Scheme, I had thought this was an optional step so I did not define this in the ApolloServerSettingsMandatory.properties file. If we had multiple connectors say 5 connectors do we just copy the following code 5 times:
Gateway.External.SchemaResource=
Gateway.External.ChartingSchemesResource=

Cheers

**Comment by Anthony-Johnson-i2** [comment]
Hi Raymond

The available port numbers will largely depend on what else you have already using ports on your system or what ports  are allowed though any firewalls or anything that is set up.

A list of ports and what they are used for is maintained   by IANA (Internet Assigned Numbers Authority)

https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml

Ports between 0 and 1023 are system ports so should be avoided  and ports 49152 to 65535 get dynamically allocated so should be avoided as well for your own applications.

So usually you would chose a port number between 1024 and 49151  that is not already in use and make it something that is easy to remember if  that helps.

To see what ports are currently in use on your windows machine you can open a command prompt as an administrator and type  netstat -ano  (you can type netstat /h for a full list of options)  or you can use an application like Process Monitor from Microsoft.

**Comment by Anthony-Johnson-i2** [comment]
Hi Raymond, a Labelling scheme is required but as it defined as part of the schema  file  and there are a number of ways that this schema or schemas can be referenced.

You can configure things so that your connector uses a gateway schema or  provides its own schema  or uses the schema from the Information store  if you are using one. 

If you have multiple connectors then they can all use the same schema defined as a gateway  schema or can use different schemas defined as gateway schemas or can again use connector schemas they provide themselves or can use the information store schema if you have an information store , it all depends on what works best for your situation.

The gateway schema option can sometimes be a bit misleading as

```
Gateway.External.SchemaResource=
Gateway.External.ChartingSchemesResource=
```

are not statically named properties like the others are,  and as you have guessed you can have more than one pair of them in the file. 

The names of these properties contain a short name for the  the gateway schema which is just a way of linking them up to the connectors that use them in this case it is ` External` and you can do that with extra configuration in the topology file or in the connector config file.

So if you had two connectors and you wanted them to use two different gateway schemas you might have something like this  defined in `ApolloServerSettingsMandatory.properties`

```
Gateway.External.SchemaResource=
Gateway.External.ChartingSchemesResource=
Gateway.Telecom.SchemaResource=
Gateway.Telecom.ChartingSchemesResource=

```
This link explains how to set that up,

https://i2group.github.io/analyze-connect/content/schemas/gateway-schema.html

The default example connector you get in the toolkit, is not set up to use a gateway schema though, it uses its own mini schema that has its own labelling scheme as part  of it :

C:\i2Group\i2analyze\toolkit\examples\connectors\example-connector\schema.xml


![image](https://user-images.githubusercontent.com/98407720/183446225-e1d11c4a-39a6-4e16-9565-b64054bfa1b4.png)

So I expect you will need to check through  to see if you have altered any of its default set up.

Cheers

**Comment by RaymondVA** [comment]
Hi Anthony,

The ports explanation is very clear so thank you!

When you said "You can configure things so that your connector uses a gateway schema or **provides its own schema**", and because we always need to define a schema for a connector, this means I can just leave the schema with the connector files (and I don't need to create a gateway schema)? I just find the gateway schema a bit redundant, so I wouldn't want to implement it. But I also understand its benefits to test/view my schema on ANBP but this is the only value I think it provides?

Now I think I understand better that the error I had was asking for a labelling schema in the example schema.xml file. Do we always need to ensure that we have defined a labelling schema when creating the schema for the connector?
